lib/openssl/asn1.rb

```diff
@@ -150,30 +278,317 @@ def each(&blk)
+
+    class ObjectId < Primitive
+      # def der_value
+      #   value = oid.split(".").map(&:to_i)
+
+      #   return (40 * value[0]).chr if value.length == 1
+
+      #   data = "".b
+      #   data << (40 * value[0] + value[1]).chr
+
+      #   rest = data[2..-1]
+      #   return data unless rest
+
+      #   rest.each do |v|
+      #     if v < 0x80
+      #       data << v.chr
+      #     else
+      #       octets = "".b
+      #       n = v
+      #       begin
+      #         octets.prepend((n & 0x7f | 0x80).chr)
+      #         n = n >> 7
+      #       end until n == 0
+      #       octets[-1] = (octets[-1].ord & 0x7f).chr
+      #       data << octets
+      #     end
+      #   end
+
+      #   data
+      # end
+    end
+
+    class UTCTime < Primitive
+    end
+
+    class GeneralizedTime < Primitive
     end
 
     class EndOfContent < ASN1Data
       def initialize
         super("", 0, :UNIVERSAL)
       end
+
+      def to_der
+        "\x00\x00".b
+      end
+    end
+
+    class Set < Constructive
+    end
+
+    class Sequence < Constructive
+
+    end
+
+    module_function
+
+    # ruby port of ASN1_object_size
+    def object_size(indefinite_length, length, tag)
+      ret = 1
+
+      return -1 if length < 0
+
+      if tag >= 31
+        while tag > 0
+          tag >>= 7
+          ret += 1
+        end
+      end
+      if indefinite_length
+        ret += 3
+      else
+        ret += 1
+        if length > 127
+          tmplen = length
+          while tmplen > 0
+            tmplen >>= 8
+            ret+=1
+          end
+        end
+      end
+
+      return -1 if (ret >= INT_MAX - length)
+
+
+      ret + length
+    end
+
+    # ruby port of openssl ASN1_put_object
+    def put_object(constructed, indefinite_length, length, tag, tag_class)
+      str = "".b
+      xclass = take_asn1_tag_class(tag_class)
+
+      i = constructed ? V_ASN1_CONSTRUCTED : 0
+      i |= (xclass & V_ASN1_PRIVATE)
+
+      if tag < 31
+        str << (i | (tag & V_ASN1_PRIMITIVE_TAG)).chr
+
+      else
+        str << (i | V_ASN1_PRIMITIVE_TAG).chr
+
+        i = 0
+        ttag = tag
+
+        while ttag > 0
+          i += 1
+          ttag >>= 7
+        end
+
+        ttag = i
+
+        while i > 0
+          i -= 1
+          tag_str = tag & 0x7f
+          if (i != (ttag - 1))
+            tag_str |= 0x80
+          end
+          str.insert(1, tag_str.chr)
+          tag >>= 7
+        end
+      end
+
+      if constructed && indefinite_length
+        str << 0x80.chr
+      else
+        str << put_length(length)
+      end
+      str
+    end
+
+
+    def put_length(length)
+      raise ASN1Error, "invalid length" if length < 0
+
+      if length < 0x80
+        length.chr
+      else
+        data = integer_to_octets(length)
+        (data.size | 0x80).chr << data
+      end
+    end
+
+    def put_integer(value)
+      raise TypeError, "Can't convert nil into Boolean" if value.nil?
+
+      value = value.to_i
```

Comment:
I wonder if there is a way to utilize `OpenSSL::BN#to_s` (or maybe it's worth adding a private helper method in C).

While speed is not the top priority this time, INTEGER often carries a huge number and looping on every byte can be too slow.

Reply:
I guess we need to benchmark to access the regression here.

Reply:
I thought more about this and, perhaps one could move the C implementation to a `BN#to_der` method, and call it here, what do you think?

Reply:
The current Ruby implementation takes super-linear time due to `Integer#>>` allocating Bignums within loop:

```
require"openssl"
require"benchmark"
def d(bit,sig) asn1=OpenSSL::ASN1::Integer(OpenSSL::BN.rand(bit)*sig); 10_000.times{asn1.to_der} end
Benchmark.bm do |a|
  a.report(2048){d(2048,1)}
  a.report(4096){d(4096,1)}
  a.report(8192){d(8192,1)}
  a.report(-2048){d(2048,-1)}
  a.report(-4096){d(4096,-1)}
  a.report(-8192){d(8192,-1)}
end

# master, C impl
       user     system      total        real
2048  0.007704   0.000886   0.008590 (  0.008631)
4096  0.009238   0.003092   0.012330 (  0.012395)
8192  0.018515   0.000831   0.019346 (  0.019458)
-2048  0.006451   0.000114   0.006565 (  0.006601)
-4096  0.009743   0.000000   0.009743 (  0.009776)
-8192  0.017901   0.000000   0.017901 (  0.017972)

# Current Ruby impl
       user     system      total        real
2048  0.902110   0.006856   0.908966 (  0.913044)
4096  2.042627   0.015922   2.058549 (  2.066914)
8192  4.905958   0.017925   4.923883 (  4.944419)
-2048  1.324276   0.002986   1.327262 (  1.332936)
-4096  3.018007   0.011945   3.029952 (  3.043331)
-8192  6.934900   0.032854   6.967754 (  6.997431)
```

I did experiment with `bn.to_s(2)`. This takes linear time, and while positive numbers are fast enough, negative numbers are still much slower than the C impl.

```ruby
def put_integer(value)
  raise TypeError, "Can't convert nil into OpenSSL::BN" if value.nil?

  value = value.to_bn
  if value >= 0
    data = value.to_s(2)
    data.prepend("\x00".b) if data.empty? || data.getbyte(0) >= 0x80
  else
    data = (-value - 1).to_s(2).bytes.map { |b| b ^ 0xff }.pack("C*")
    data.prepend("\xff".b) if data.empty? || data.getbyte(0) < 0x80
  end

  data
end

# New Ruby impl
       user     system      total        real
2048  0.014312   0.000040   0.014352 (  0.014420)
4096  0.016308   0.000000   0.016308 (  0.016384)
8192  0.021418   0.000000   0.021418 (  0.021499)
-2048  0.151956   0.000000   0.151956 (  0.152512)
-4096  0.285909   0.000892   0.286801 (  0.287931)
-8192  0.546144   0.001994   0.548138 (  0.550484)
```

Reply:
`BN#to_der` would not be appropriate since BN is just a number and we can't define a DER encoding for it. We should use something else if we add a C implementation.

Reply:
```ruby
def put_integer(value)
  raise TypeError, "Can't convert nil into OpenSSL::BN" if value.nil?

  value = value.to_bn
  if value >= 0
    data = value.to_s(2)
    data.prepend("\x00".b) if data.empty? || data.getbyte(0) >= 0x80
  else
    value = (1.to_bn << (value.num_bits + 7) / 8 * 8) + value
    data = value.to_s(2)
    data.prepend("\xff".b) if data.empty? || data.getbyte(0) < 0x80
  end

  data
end
```

Revised it to avoid iterating over Array:

```
       user     system      total        real
2048  0.015041   0.002001   0.017042 (  0.017112)
4096  0.019514   0.001055   0.020569 (  0.020839)
8192  0.023327   0.000000   0.023327 (  0.023450)
-2048  0.019398   0.000000   0.019398 (  0.019505)
-4096  0.023691   0.000000   0.023691 (  0.023826)
-8192  0.030820   0.000000   0.030820 (  0.030999)
```

This speed looks acceptable for now and maybe it's not worth adding a C helper method.

Reply:
that's pretty good indeed. I'd expect YJIT to improve on those, if it's worth benchmarking. Added your implementation.